feat: mgr MIP21 integration

xmxanuel opened the pull request:
This PR makes the `mgr` contract compatible with MIP21.

- mgr acts as MIP21 operator and output conduit at the same time
- underlying collateral would be the RWA token instead of DROP
- DROP is kept in the mgr
- MKR governance can trigger a soft liquidation in Tinlake 
- mgr contract doesn't need any permissions in dss directly only on the rwaUrn (MIP21)

xmxanuel committed:
fix: compiler version

xmxanuel committed:
dapp install https://github.com/livnev/rwa-example

ilinzweilin committed:
remove dss

ilinzweilin committed:
remove dss

ilinzweilin committed:
remove dss-interfaces

ilinzweilin committed:
remove dss-interfaces

ilinzweilin committed:
fix submodules

xmxanuel committed:
added mgr mip21 draft

xmxanuel committed:
add comments

MrChico commented on src/mgr.sol:
```diff
@@ -128,43 +131,43 @@ contract TinlakeManager {
     RedeemLike   public pool;
 
+    // MIP21 RWAUrn
+    MIP21UrnLike public urn;
```
where is this given a value?

xmxanuel replied:
In the `file` method.
 Because we have bi-directional dependency with RwaUrn if the mgr is the operator and output-conduit

MrChico replied:
only in file I suppose

MrChico commented on src/mgr.sol:
```diff
@@ -186,124 +189,117 @@ contract TinlakeManager {
     }
 
     function wipe(uint256 wad) public operatorOnly {
         require(safe && live, "TinlakeManager/bad-state");
```
these checks are kind of redundant since anybody can call `urn.wipe`

MrChico commented on src/mgr.sol:
```diff
@@ -128,43 +131,43 @@ contract TinlakeManager {
-        vat.hope(daiJoin_);
-
+        // rwaUrn allowance for rwaToken
+        rwaToken.approve(address(urn), uint(-1));
```
this needs to happen in `file` as well?

xmxanuel replied:
We could make the approval in the `lock` method for the specific amount

MrChico replied:
I think it's fine to move to `file`.

lucasvo replied:
Putting it into file is riskier because it requires `urn` to be set which it might not be at the time. Given `lock` is only executed once it doesn't add any inefficiency.

MrChico commented on src/mgr.sol:
```diff
@@ -186,124 +189,117 @@ contract TinlakeManager {
 
-    // --- Writeoff ---
+    // --- Write-off ---
+    // method should be called before RwaLiquidationOracle.cull()
```
not sure what purpose this function plays anymore

MrChico replied:
only to determine the `tab` at time of liquidation I suppose.

xmxanuel replied:
Yes, that currently the only purpose. Do you see a more elegant way?

lucasvo replied:
There was some discussion around this in other channels. In short: Without touching MIP21 we can't make it anymore elegant.

MrChico commented on src/mgr.sol:
```diff
@@ -186,124 +189,117 @@ contract TinlakeManager {
+
+    // moves the rwaToken into the vault
+    // requires that mgr contract holds the rwaToken
+    function lock(uint wad) public operatorOnly {
```
should we simply call `urn.lock` in `join`?

xmxanuel replied:
It depends. If we want to lock 100% of the rwaToken in the beginning or calculated the percentage of locked DROP (required drop to borrow up to the the debt ceiling would be 100%) and move the corresponding rwaToken amount

godsflaw commented on src/mgr.sol:
```diff
@@ -16,7 +16,7 @@
 // along with this program.  If not, see <https://www.gnu.org/licenses/>.
 
-pragma solidity 0.5.12;
+pragma solidity >=0.5.15 <0.6.0;
```
`0.5.12` or `0.6.12`

MrChico commented on src/mgr.sol:
```diff
@@ -128,43 +131,43 @@ contract TinlakeManager {
-        dai.approve(daiJoin_, uint256(-1));
-        vat.hope(daiJoin_);
-
+        // rwaUrn allowance for rwaToken
+        rwaToken.approve(address(urn), uint(-1));
```
```suggestion
```

MrChico commented on src/mgr.sol:
```diff
@@ -186,124 +189,117 @@ contract TinlakeManager {
         else if (what == "end")  end = EndLike(data);
+        else if (what == "urn") {
+            urn = MIP21UrnLike(data);
+            dai.approve(data, uint256(-1));
```
```suggestion
            dai.approve(data, uint256(-1));
            // rwaUrn allowance for rwaToken
            rwaToken.approve(address(urn), uint(-1));
```

lucasvo replied:
Not necessary anymore. We do it in `lock()`

godsflaw commented on src/mgr.sol:
```diff
@@ -186,124 +189,117 @@ contract TinlakeManager {
         emit Draw(wad);
     }
 
     function wipe(uint256 wad) public operatorOnly {
```
think about this with regards to operator going away.

godsflaw commented on src/mgr.sol:
```diff
@@ -186,124 +189,117 @@ contract TinlakeManager {
-        vat.grab(ilk, address(this), address(this), address(this),
-                 -int256(dropReturned), 0);
-        vat.slip(ilk, address(this), -int256(dropReturned));
+        urn.wipe(payBack);
```
check MIP21 in ES.

lucasvo commented on src/mgr.sol:
```diff
@@ -128,43 +131,43 @@ contract TinlakeManager {
-        vat.hope(daiJoin_);
-
+        // rwaUrn allowance for rwaToken
+        rwaToken.approve(address(urn), uint(-1));
```
```suggestion
        rwaToken.approve(address(urn), uint256(-1));
```

lucasvo replied:
Also review other occurrences of uint256

lucasvo committed:
Refactored auth: `operators` are now wards and maker governance is now
`governance`
uint256 consistency
require message consistency (`TinlakeMgr`)
rwaToken & urn now only set via file

lucasvo committed:
Added owner to know where to return dai to

lucasvo committed:
comment on migrate

lucasvo committed:
Work on tests

lucasvo committed:
Test for lock

lucasvo committed:
Initial tests for mgr

ilinzweilin committed:
migrate unit tests

ilinzweilin committed:
add unit tests for unwind

lucasvo committed:
Merge pull request #28 from centrifuge/unit-tests

migrate unit tests

lucasvo committed:
Update src/mgr.sol

lucasvo committed:
Merge pull request #27 from centrifuge/auth-mod

Auth Improvements & tests

lucasvo committed:
Update rwa-example

lucasvo committed:
Updating auth structure

lucasvo committed:
Uint256

lucasvo committed:
Merge pull request #29 from centrifuge/fix-rwaexample

Update rwa-example

lucasvo committed:
Fix solc version in actions

ilinzweilin committed:
fix build: downgrade compiler version

lucasvo committed:
Merge pull request #30 from centrifuge/fix-build

fix build: downgrade compiler version

lucasvo committed:
Removing unnecessary live check

godsflaw commented on src/mgr.sol:
```diff
@@ -16,7 +16,9 @@
-pragma solidity 0.5.12;
+pragma solidity >=0.5.12 <0.6.0;
+
+import "ds-test/test.sol";
```
Did you forget to remove this?

godsflaw commented on src/mgr.sol:
```diff
@@ -16,7 +16,9 @@
 // along with this program.  If not, see <https://www.gnu.org/licenses/>.
 
-pragma solidity 0.5.12;
+pragma solidity >=0.5.12 <0.6.0;
```
no harm in pegging this to `0.5.12`

godsflaw commented on src/mgr.sol:
```diff
@@ -87,30 +94,39 @@ contract TinlakeManager {
+    modifier governance {
+        require(can[msg.sender] == 1, "TinlakeMgr/not-governance");
         _;
     }
```
Governance should not need access to the manager. The RwaLiquidationOracle can be checked to change manager states.

```suggestion

```

godsflaw commented on src/mgr.sol:
```diff
@@ -87,30 +94,39 @@ contract TinlakeManager {
     event Rely(address indexed usr);
     event Deny(address indexed usr);
+    event Hope(address indexed who);
+    event Nope(address indexed who);
```
```suggestion
```

godsflaw commented on src/mgr.sol:
```diff
@@ -71,14 +76,16 @@ interface RedeemLike {
+    // Wards are for Centrifuge
     mapping (address => uint256) public wards;
+    // can is Maker governance
+    mapping (address => uint256) public can;
```
```suggestion
```

godsflaw commented on src/mgr.sol:
```diff
@@ -186,128 +205,131 @@ contract TinlakeManager {
+    // a redeemOrder is submitted to receive DAI back
+    function tell() public governance {
+        require(safe, "TinlakeMgr/not-safe");
+        uint256 ink = gem.balanceOf(address(this));
```
Remove the modifier for `governance` and use the `RwaLiquidationOracle` status to make the updates in this function.

godsflaw commented on src/mgr.sol:
```diff
@@ -186,128 +205,131 @@ contract TinlakeManager {
-        require(art <= 2 ** 255, "TinlakeManager/overflow");
+    // --- Write-off ---
+    // method should be called before RwaLiquidationOracle.cull()
+    function sink() public governance {
```
We can also remove `governance` here, and check the `RwaLiquidationOracle` status to update the `tab`.

lucasvo replied:
I don't think there's a way to get the tab from the `RwaLiquidationOracle`. Am I missing something?

lucasvo replied:
This would be the minimal change we need I believe to make it work:

```solidity
contract Liq is RwaLiquidationOracle {
    uint public tab;
    constructor (address vat_, address vow_) RwaLiquidationOracle(vat_, vow_) public{}
    function cull(bytes32 ilk, address urn) external auth {
        (uint256 ink, uint256 art) = vat.urns(ilk, urn);
        tab = mul(rate, art);
        super.cull(ilk, urn);
    }
}
```

godsflaw commented on src/mgr.sol:
```diff
@@ -128,42 +144,45 @@ contract TinlakeManager {
-                address end_,      bytes32 ilk_
+    constructor(address dai_,       address daiJoin_,
+                address drop_,      address pool_,
+                address governance, address owner_,
```
Remove `governance` and add the `urn_` and `oracle_` here.

godsflaw commented on src/mgr.sol:
```diff
@@ -128,42 +144,45 @@ contract TinlakeManager {
-        operator = operator_;
+
+        can[governance] = 1;
+        emit Hope(governance);
```
```suggestion
```

godsflaw commented on src/mgr.sol:
```diff
@@ -128,42 +144,45 @@ contract TinlakeManager {
-        vow = vow_;
-
+        vat = VatLike(vat_);
+        end = EndLike(end_);
```
```suggestion
        end = EndLike(end_);
        urn = MIP21UrnLike(urn_);
        oracle = MIP21OracleLike(oracle_);
```

godsflaw commented on src/mgr.sol:
```diff
@@ -128,42 +144,45 @@ contract TinlakeManager {
     RedeemLike   public pool;
 
+    // MIP21 RWAUrn
+    MIP21UrnLike public urn;
```
```suggestion
    MIP21UrnLike public urn;
    MIP21OracleLike public oracle;
```

godsflaw commented on src/mgr.sol:
```diff
@@ -186,128 +205,131 @@ contract TinlakeManager {
-        operator = newOperator;
-        emit SetOperator(newOperator);
+    function free(uint256 wad) public auth {
+        urn.quit();
```
I think the DAI coming back from the `quit()` will get stuck here.  You may want to check `dai.balanceOf(address(this))` and forward that on to the appropriate/recoverable place.

lucasvo replied:
In that case we would actually deal with it using `migrate` since we're not quite sure what we should do with this DAI after ES anyway. I think this is theoretically not a case that should happen with our use of MIP21 anyway - we wouldn't have any DAI accrue in the Urn.

Never the less, I renamed this method to `quit` to make it a bit more intuitive and actually also implemented a `free` that takes out the RWAToken.

godsflaw commented on src/mgr.sol:
```diff
@@ -186,128 +205,131 @@ contract TinlakeManager {
-        vat.hope(dst);
+    // --- Administration ---
+    function migrate(address dst) public {
+        require(can[msg.sender] == 1 || wards[msg.sender] == 1, "TinlakeMgr/auth-or-gov-only");
```
Since we don't need governance, we can remove this.  Our only suggestion is to ensure this migrates your manager correctly.
```suggestion
    function migrate(address dst) public auth {
```

godsflaw commented on src/mgr.sol:
```diff
@@ -186,128 +205,131 @@ contract TinlakeManager {
+          ilk = what;
+        } else revert("TinlakeMgr/file-unknown-param");
+    }
+
```
We shouldn't need to file another ilk
```suggestion
```

godsflaw commented on src/mgr.sol:
```diff
@@ -128,42 +144,45 @@ contract TinlakeManager {
+                address drop_,      address pool_,
+                address governance, address owner_,
+                address tranche_,   address end_,
+                address vat_,       bytes32 ilk_
```
perhaps we don't need an ilk if we have the `urn.gemJoin().ilk()`?

lucasvo replied:
Yes, removed.

godsflaw commented on src/mgr.sol:
```diff
@@ -186,128 +205,131 @@ contract TinlakeManager {
-        else if (what == "daiJoin") daiJoin = JoinLike(data);
-        else if (what == "end")  end = EndLike(data);
-        else revert("Vat/file-unrecognized-param");
+        if (what == "urn") {
+            urn = MIP21UrnLike(data);
+            dai.approve(data, uint256(-1));
+        }
```
You probably don't need to do this either.  The `urn` shouldn't change.
```suggestion
        if (what == "urn") {
            urn = MIP21UrnLike(data);
            dai.approve(data, uint256(-1));
        }
```

godsflaw commented on src/mgr.sol:
```diff
@@ -186,128 +205,131 @@ contract TinlakeManager {
+        else if (what == "owner") {
+            owner = data;
+        }
+        else if (what == "rwaToken") {
+            rwaToken = GemLike(data);
+        }
```
This is also immutable.  Also, you can get this from the `urn.gemJoin().gem() `.
```suggestion
```

lucasvo replied:
Nice!

godsflaw reviewed (changes requested):
We didn't get through the entire thing, but this is looking good.  You may want to lift the deploy script from the `rwa-example` repo and adapt it to deploy the manager as the output/input conduit.

ilinzweilin committed:
add: recover tests & fixes

ilinzweilin committed:
fix: merge MIP21_Integration

ilinzweilin committed:
fix: ds-test import

lucasvo committed:
Using the RWALiquidationOracle to get tab (Dai owed to maker)

xmxanuel commented on src/mgr.sol:
```diff
@@ -186,128 +166,140 @@ contract TinlakeManager {
-        vat.grab(ilk, address(this), address(this), address(this),
-                 -int256(dropReturned), 0);
-        vat.slip(ilk, address(this), -int256(dropReturned));
+        tab = mul(art, rate);
```
I think the actual update of  `tab` should happen after the `urn.wipe`. It would currently not include the `disburse redeemedDAI`. If I understand correctly

xmxanuel committed:
added urn check in cull

xmxanuel committed:
fix: testcase for testUnwindPartialRepayment

xmxanuel committed:
local variable for tab in unwind

ilinzweilin committed:
fix merge conflicts

xmxanuel committed:
added pool and tranche to file method

ilinzweilin committed:
fix from pr review

xmxanuel committed:
Merge pull request #32 from manuelpolzhofer/fixMIP21Unwind

fix: mip21 require liq.cull call before mgr.cull

ilinzweilin committed:
fix merge conflicts

ilinzweilin committed:
Merge pull request #31 from centrifuge/unit-tests-recover

add: recover tests & fixes

lucasvo committed:
Remove unused mocks

godsflaw reviewed (approved):
The Maker SC Domain Team believes this PR should allow the manager to function properly as a `MIP21` style conduit; however, we offer no warranty -- implied or otherwise -- as to the correctness of this PR.  It should be understood that anything beyond the MIP21 core code is treated as a blackbox by our team, and that it's the responsibility of this code's authors to ensure correct functioning and operation of the manager.  That being said, from what review we did perform, we have confidence in the team's capabilities, given that the requisite appropriate care and attention are given, to ensure the proper functioning of this module.
